src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -25,29 +24,73 @@ internal sealed partial class CSharpSelectionValidator(
     {
         private readonly bool _localFunction = localFunction;
 
-        protected override SelectionInfo UpdateSelectionInfo(
-            SelectionInfo selectionInfo, StatementSyntax? firstStatement, StatementSyntax? lastStatement, CancellationToken cancellationToken)
+        protected override InitialSelectionInfo GetInitialSelectionInfo(CancellationToken cancellationToken)
+        {
+            var root = this.SemanticDocument.Root;
+
+            var adjustedSpan = GetAdjustedSpan(OriginalSpan);
+
+            var firstTokenInSelection = root.FindTokenOnRightOfPosition(adjustedSpan.Start, includeSkipped: false);
+            var lastTokenInSelection = root.FindTokenOnLeftOfPosition(adjustedSpan.End, includeSkipped: false);
+
+            if (firstTokenInSelection.Kind() == SyntaxKind.None || lastTokenInSelection.Kind() == SyntaxKind.None)
+                return InitialSelectionInfo.Failure(FeaturesResources.Invalid_selection);
+
+            var commonRoot = firstTokenInSelection.GetCommonRoot(lastTokenInSelection);
+            var selectionInExpression = commonRoot is ExpressionSyntax;
+
+            var statusReason = CheckSpan();
+            if (statusReason is not null)
+                return InitialSelectionInfo.Failure(statusReason);
+
+            return CreateInitialSelectionInfo(
+                selectionInExpression, firstTokenInSelection, lastTokenInSelection, cancellationToken);
+
+            string? CheckSpan()
+            {
+                if (firstTokenInSelection.SpanStart > lastTokenInSelection.Span.End)
+                    return FeaturesResources.Selection_does_not_contain_a_valid_token;
+
+                if (!UnderValidContext(firstTokenInSelection) || !UnderValidContext(lastTokenInSelection))
+                    return FeaturesResources.No_valid_selection_to_perform_extraction;
+
+                if (commonRoot == null)
+                    return FeaturesResources.No_common_root_node_for_extraction;
+
+                if (!commonRoot.ContainedInValidType())
+                    return FeaturesResources.Selection_not_contained_inside_a_type;
+
+                if (!selectionInExpression && !commonRoot.UnderValidContext())
+                    return FeaturesResources.No_valid_selection_to_perform_extraction;
+
+                return null;
+            }
+        }
+
+        protected override FinalSelectionInfo UpdateSelectionInfo(
+            InitialSelectionInfo initialSelectionInfo,
```

Comment:
initialselection is now passed in data.  FinalSelection is what is passed out.  the domains are kept separated.